pilot/pkg/controller/workloadentry/workloadentry_controller.go

```diff
@@ -243,54 +244,76 @@ func (c *Controller) processNextWorkItem() bool {
 		return true
 	}
 
-	err := c.unregisterWorkload(workItem.entryName, workItem.proxy, workItem.disConTime, workItem.origConTime)
+	err := c.unregisterWorkload(workItem.entryName, workItem.proxy, workItem.disConTime, workItem.origConTime, workItem.autoCreated)
 	c.handleErr(err, item)
 	return true
 }
 
 func setConnectMeta(c *config.Config, controller string, conTime time.Time) {
+	if c.Annotations == nil {
+		c.Annotations = make(map[string]string)
+	}
 	c.Annotations[WorkloadControllerAnnotation] = controller
 	c.Annotations[ConnectedAtAnnotation] = conTime.Format(timeFormat)
 	delete(c.Annotations, DisconnectedAtAnnotation)
 }
 
 func (c *Controller) RegisterWorkload(proxy *model.Proxy, conTime time.Time) error {
-	if !features.WorkloadEntryAutoRegistration || c == nil {
+	if c == nil {
 		return nil
 	}
-	// check if the WE already exists, update the status
-	entryName := autoregisteredWorkloadEntryName(proxy)
+	var entryName string
+	var autoCreate bool
+	if features.WorkloadEntryAutoRegistration &&
+		proxy.Metadata.AutoRegisterGroup != "" {
+		entryName = autoregisteredWorkloadEntryName(proxy)
+		autoCreate = true
+	} else if features.WorkloadEntryHealthChecks &&
+		proxy.Metadata.ProxyConfig != nil && proxy.Metadata.ProxyConfig.ReadinessProbe != nil {
+		entryName = lookupWorkloadEntryByProxyID(c.store, proxy)
+	}
 	if entryName == "" {
 		return nil
 	}
-	proxy.AutoregisteredWorkloadEntryName = entryName
+	proxy.WorkloadEntryName = entryName
+	proxy.WorkloadEntryAutoCreated = autoCreate
 
 	c.mutex.Lock()
 	c.adsConnections[makeProxyKey(proxy)]++
 	c.mutex.Unlock()
 
-	if err := c.registerWorkload(entryName, proxy, conTime); err != nil {
+	err := c.connectWorkload(entryName, proxy, conTime, autoCreate)
+	if err != nil {
 		log.Errorf(err)
+	}
+	return err
+}
+
+func (c *Controller) connectWorkload(entryName string, proxy *model.Proxy, conTime time.Time, autoCreate bool) error {
```

Comment:
add an comment what does this function do

Reply:
added comments to this and other functions affected by this PR